### DISCUSSION TITLE
Expose dialect in engine

### DISCUSSION THREAD

**Discussion by holymonson** [sqlalchemy/sqlalchemy#5974] [Usage Questions]
**Is your feature request related to a problem? Please describe.**

Take an postgresql engine as example, `engine.dialect` is binded to a `sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2` object.
And there is no attribute in `engine` nor `PGDialect_psycopg2` referring to module `sqlalchemy.dialects.postgresql`.
I'm trying to refer to `sqlalchemy.dialects.postgresql.insert` from `engine` but fail.

In my understanding (from https://docs.sqlalchemy.org/en/13/dialects/), `postgresql` is a __dialect__ while `psycopg2` is its __driver__, so `engine.dialect` pointing to a driver confuses me a little.

**Describe the solution you'd like**

Some way to refer to `sqlalchemy.dialects.postgresql` from `engine`. And it would be nice to have some clarification of those terms.

**Describe alternatives you've considered**

dml on a table binding to an engine should return its dialect version.
For example, `table.insert()` will return a `sqlalchemy.sql.dml.Insert` object, but if the table has already binded to a postgresql engine, then `table.insert()` should return a `sqlalchemy.dialects.postgresql.dml.Insert` object.

**Have a nice day!**

**Comment by CaselIT** [comment]
Hi,

Not a full reply, just some clarifications (hopefully).

> postgresql is a dialect while psycopg2 is its driver, so engine.dialect pointing to a driver confuses me a little.

`engine.dialect` is not pointing to the driver, but to a sqlalchemy dialect. `PGDialect_psycopg2` is a subclass of the general postgres dialect `PGDialect` specific for psycopg2, but it's not the psycopg2 package.

Also `sqlalchemy.dialects.postgresql.insert` is a function, while `sqlalchemy.dialects.postgresql` is a package, so I'm not sure how you could refer them from the engine.

Regarding the table, the whole concept of binding tables etc is deprecated (in 1.4 iirc) and discouraged.
In general the table (or any construct) would have no way of knowing that it's executing against pg or another db

**Comment by zzzeek** [comment]
The ``engine.dialect`` attribute points to the dialect.   ``sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2`` is a dialect even though "psycopg2" is in its name, that's the dialect that both knows about PostgreSQL and also knows how to interpret it in the way that psycopg2 wants.

to get to the psycopg2 module itself you would use ``engine.dialect.module``.   But this is not a common use case as if someone has DBAPI-specific code, they should just use "import psycopg2".

Similarly, if you want to use "sqlalchemy.dialects.postgresql.insert", that implies you have code that is explicit to this object, so again, you should import it:

    from sqlalchemy.dialects.postgresql import insert

if you didn't have code that's explicit to this object then you wouldn't need to import it, you can use the plain "insert".

I'm going to take a guess that what you **really** want is some way to use the Postgresql or SQLite versions of these consructs, as they are very similar?  if so, one way to approach that is like this:

    from sqlalchemy.dialects.postgresql import insert as postgresql_insert
    from sqlalchemy.dialects.sqlite import insert as sqlite_insert

    def insert_for_db(engine):
       if engine.name == "postgresql":
           return postgresql_insert
       elif engine.name == "sqlite":
           return sqlite_insert
       else:
           assert False

or you could use ``__import__``:

    mod = getattr(__import__(f"sqlalchemy.dialects.{engine.name}").dialects, engine.name)
    insert = mod.insert
